src/SignalR/common/Shared/AckPipeReader.cs

```diff
@@ -0,0 +1,187 @@
+                // While they technically point to different positions, they point to the same concept of "beginning of the next buffer"
+                var ackSlice = buffer.Slice(_ackDiff);
+                if (buffer.Slice(_consumed).First.Length == 0 && ackSlice.Start.GetInteger() == 0)
+                {
+                    // Fix consumed to point to the beginning of the next Segment
+                    _consumed = ackSlice.Start;
+                }
+                else if (!_consumed.Equals(default))
+                {
+                    var consumedLength = buffer.Slice(_consumed).Length;
+                    if (consumedLength == ackSlice.Length)
+                    {
+                        _consumed = default;
+                    }
+                    else if (consumedLength > ackSlice.Length)
+                    {
+                        // ack is greater than consumed, should not be possible
+
+                        // TODO: verify that if ack is less than total but more than consumed this isn't hit
+                        // e.g. 13 bytes in underlying pipe, only consumed 11 during Read+Advance. Will an ack id of 12 be allowed?
+                        Debug.Assert(false);
+                    }
+                    else if (consumedLength < ackSlice.Length)
+                    {
+                        // this is normal, ack id is less than total written
+                    }
+                }
+
+                buffer = ackSlice;
+                _ackId += _ackDiff;
+                _ackDiff = 0;
+                _ackPosition = buffer.Start;
+            }
+        }
+
+        // Slice consumed, unless resending, then slice to ackPosition
+        if (_resend)
+        {
+            _resend = false;
+            if (buffer.Length != 0 && !_ackPosition.Equals(default))
+            {
+                buffer = buffer.Slice(_ackPosition);
+            }
+            // update total written if there is more written to the pipe during a reconnect
+            // TODO: add tests for both these paths
+            if (!_consumed.Equals(default))
+            {
+                Debug.Assert(buffer.Length - buffer.Slice(_consumed).Length >= 0);
+                _totalWritten += buffer.Length - buffer.Slice(_consumed).Length;
+            }
+            else
+            {
+                _totalWritten += buffer.Length;
+            }
+        }
+        else if (buffer.Length > 0)
+        {
+            _ackPosition = buffer.Start;
+            if (!_consumed.Equals(default))
+            {
+                buffer = buffer.Slice(_consumed);
+            }
+            _totalWritten += (uint)buffer.Length;
+        }
+
+        res = new(buffer, res.IsCanceled, res.IsCompleted);
+        return res;
+    }
+
+    public override bool TryRead(out ReadResult result)
+    {
+        throw new NotImplementedException();
```

Comment:
is this just temporary?

Reply:
Correct. Since the main use case is SignalR and SignalR doesn't call `TryRead` this isn't needed in the first iteration. For completeness it should be implemented (and it looks like it'd be easy too).